### ISSUE TITLE
sql: support subqueries over tables in UPDATE...SET clauses

### ISSUE THREAD

**Issue by sploiselle** [MaterializeInc/materialize#9274]
Right now, `UPDATE...SET` disallows subqueries as a very strict prevention of reading values that could introduce serializability violations.

However, I think the restriction actually needs to be only that the subqueries can reference other tables (i.e. within the same time domain).

e.g. this should be permitted:

```
CREATE TABLE u (a int, b int);
CREATE TABLE m (a int);
INSERT INTO u VALUES (1,2), (3, 4);
INSERT INTO m VALUES (5);
UPDATE u SET b = (SELECT max(a) FROM m);
SELECT * FROM u;
```
```
 a | b 
---+---
 1 | 5
 3 | 5
```

The write lock we currently hold for the lifetime of `ReadThenWrite` plans guarantees that any table read from during its execution will not change.

@mjibson does this check out to you, as well?

**Comment by mjibson** [comment]
Yes, that sounds correct. Don't we already do roughly the same thing during `UPDATE .. FROM`?
